### ISSUE TITLE
Use cases

### ISSUE THREAD

**Issue by ibc** [indutny/gyp.js#19]
Given #9 it seems that this project needs to define its own targets. Here a suggested list:

1. Be a replacement for Python `gyp` (currently bundled into `node-gyp`), so `node-gyp` would include `gyp` as dependency and Python would not be needed to build Node.js native addons. 
1. Be able to produce `ninja` projects from Node.js (not just tiled to Node.js native addons), for example [bud](https://github.com/indutny/bud).
1. Be able to compile a newer version of Node.js from a older version of Node.js (?)

**Comment by indutny** [comment]
Thank you for bringing it up, @ibc !

The bigger our goals - the longer it will take us to get there. Thus, I think that it may be a good idea to start small with node.js addons and node-gyp, and eventually come to the full support of all GYP features.

This doesn't mean that gyp.js won't be usable in its current form for the most of the projects though.

**Comment by RangerMauve** [comment]
Getting rid of the python dependency in node-gyp would make a lot of people happy campers

**Comment by ibc** [comment]
Nice. I hope use-case 2 above will be implemented in the future, so I can replace Python `gyp` with `gyp.js` in [my project](https://github.com/ibc/mediasoup/tree/master/worker) :)

**Comment by kzc** [comment]
**4.** Bootstrap and build any version of node without a node binary using Duktape + duktape bindings + C++ compiler.

**Comment by saghul** [comment]
The way I see it compiling addons or Node itself are not that different, are they? Then there is the bootstraping issue.

**Comment by ibc** [comment]
I wonder whether with `gyp.js` something like [gyp_bud](https://github.com/indutny/bud/blob/master/gyp_bud) could be replaced by a pure JS API, so there is no even need to launch a child process to run `gyp.js`.

**Comment by indutny** [comment]
@ibc yeah, this would be great. On other hand, embedding it in [gypkg](https://github.com/indutny/gypkg) should be good too ;)

**Comment by ibc** [comment]
Sorry for coming back to this issue with an off-topic question:

Given that there is no activity in this project and given that GYP does not receive [support for Python 3](https://bugs.chromium.org/p/gyp/issues/detail?id=36), which is the expectation within the next 2 years? Python 2 will be deprecated in 2020. Why didn't Node move to `gyp.js` as the proper migration path? What should we expect those who also run projects that depend on GYP and don't want to also depend on Python 2 forever?

**Comment by refack** [comment]
The Node.js project is forking GYP, and the first item in the backlog is python3 support.
We already landed it for `node-gyp` https://github.com/nodejs/node-gyp/pull/1335

You can follow the progress at https://github.com/refack/GYP which might soon move to the nodejs org.

BTW I do plan on integrating gyp.js into GYP to have it cross-runtime portable.

**Comment by ibc** [comment]
Thanks for the update, @refack.
